src/index.js

```diff
@@ -1,174 +1,226 @@
    *
-   * @param {String|Array} room
-   * @param {Function} [fn]
+   * @param {String|String[]} room - the room(s) to empty
+   * @return {Promise} no data
    * @api public
    */
 
-  function empty(room, fn = noop) {
-    if (room && rooms[room]) {
-      if (!isArray(room)) clear(room)
-      else for (const rm of room) clear(rm)
+  function empty(room) {
+    if (isArray(room)) {
+      room.map(rm => clear(rm));
+      return Promise.all(room);
     }
-    setImmediate(fn, null, null)
+    return clear(room);
   }
 
   /**
    * Reset all rooms or a particular room.
    *
-   * @param {String} room
-   * @param {Function} [fn]
+   * @param {String} room - the room to clear
+   * @return {Promise} no data
    * @api public
    */
-
-  function clear(room, fn = noop) {
-    if ('function' === typeof room) {
-      rooms = {}, sids = {}
+  function clear(room) {
+    if (!room) {
+      rooms.clear();
+      sids.clear();
+      return Promise.resolve();
     }
-    for (const id in rooms[room]) {
-      delete sids[id][room]
+    const sockets = rooms.get(room);
+    if (sockets) {
+      for (const id of sockets) {
+        const socketRooms = sids.get(id);
+        socketRooms.delete(room);
+        if (socketRooms.size === 0) {
+          sids.delete(id);
+        }
+        rooms.delete(room);
+        // TODO update wild 
+        // wild.remove(room)
+      }
     }
-    delete rooms[room]
-    wild.remove(room)
-    setImmediate(fn, null, null)
+    return Promise.resolve();
   }
 
   /**
    * Broadcast a packet.
    *
-   * Options:
-   *  - `except` {Array} sids that should be excluded
-   *  - `rooms` {Array} list of rooms to broadcast to
-   *  - `method` {String} 'write' or 'send' if primus-emitter is present
-   *
-   * @param {Object} data
-   * @param {Object} opts
-   * @param {Object} clients Connected clients
+   * @param {*} data - the data to transmit
+   * @param {Object} options - options to use
+   * @param {String[]} options.except - spark IDs to exclude
+   * @param {String[]} options.rooms - list of the rooms to broadcast to
+   * @param {String} options.method - 'write' or 'send' if primus-emitter is present
+   * @param {Function} options.transformer - the transformer function to call on the data
+   * @param {Object} clients -  Connected clients
+   * @return {Promise.<Set>} set of sparkIDs transmitted to
    * @api public
    */
-
+  
   function broadcast(data, options = {}, clients) {
-    let rms = options.rooms || []
+    const rms = options.rooms || []
     if (!rms.length) {
-      return send(sids, clients, data, options)
-    }
-    for (const room of rms) {
-      let ids = rooms[room]
-      if (ids) send(ids, clients, data, options)
-      wild.match(room, key =>
-        send(rooms[key], clients, data, options)
+      const ids = new Set();
+      sids.forEach((value, key) => {
+        ids.add(key);
+      });
+      return send(ids, clients, data, options).then(result => new Set(result));
```

Comment:
the send function expects a set however we could do something like this `new Set([...sids.keys()])`